server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,296 @@
+     * the directory.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        logger.trace("rename() called {}, {}", source, dest);
+        localDirectory.rename(source, dest);
+        fileCache.remove(localDirectory.getDirectory().resolve(source));
+        cacheFile(dest);
+    }
+
+    /**
+     * Opens a stream for reading an existing file.
+     * Check whether the file is present locally or in remote and return the IndexInput accordingly
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        logger.trace("openInput() called {}", name);
+        // We aren't tracking temporary files (created via createTempOutput) currently in FileCache as these are created and then deleted
+        // within a very short span of time
+        // We will be reading them directory from the local directory
+        if (isTempFile(name)) {
+            return localDirectory.openInput(name, context);
+        }
+        // Return directly from the FileCache (via TransferManager) if complete file is present
+        Path key = localDirectory.getDirectory().resolve(name);
+        CachedIndexInput indexInput = fileCache.get(key);
+        if (indexInput != null) {
+            logger.trace("Complete file found in FileCache");
+            try {
+                return indexInput.getIndexInput().clone();
+            } finally {
+                fileCache.decRef(key);
+            }
+        }
+        // If file has been uploaded to the Remote Store, fetch it from the Remote Store in blocks via OnDemandCompositeBlockIndexInput
+        else {
+            logger.trace("Complete file not in FileCache, to be fetched in Blocks from Remote");
+            RemoteSegmentMetadata remoteSegmentMetadata = remoteDirectory.readLatestMetadataFile();
+            RemoteSegmentStoreDirectory.UploadedSegmentMetadata uploadedSegmentMetadata = remoteSegmentMetadata.getMetadata().get(name);
+            // TODO : Refactor FileInfo and OnDemandBlockSnapshotIndexInput to more generic names as they are not Remote Snapshot specific
+            BlobStoreIndexShardSnapshot.FileInfo fileInfo = new BlobStoreIndexShardSnapshot.FileInfo(
+                name,
+                new StoreFileMetadata(name, uploadedSegmentMetadata.getLength(), uploadedSegmentMetadata.getChecksum(), Version.LATEST),
+                null
+            );
+            return new OnDemandBlockSnapshotIndexInput(fileInfo, localDirectory, transferManager);
+        }
+    }
+
+    /**
+     * Closes the directory
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void close() throws IOException {
+        Arrays.stream(localDirectory.listAll()).forEach(f -> fileCache.remove(localDirectory.getDirectory().resolve(f)));
```

Comment:
Adding the previous comment here so it doesn't get missed:

>> @sohami We should just decRef the files in FileCache instead of force removal. That will any reference to those IndexInputs will not be affected.

> @rayshrey Why do we need to keep the IndexInputs in cache after the directory is closed. We shouldn't allow reads after directory is closed. Also not clearing these files led to IT failures due to open indexInputs

@sohami Agreed, but if there is any open reference to these local IndexInputs then above remove will cause failure to the consumers of those open input. Or is it guaranteed that Directory::close will be called only when no active open reference is present for the files under this directory instance ?

Reply:
I looked into this flow and today it seems like if there is a say PITcontext created for a PIT request, then it will be reference counted via `ReaderContext` and will be tracked in map of `activeReaders` at shard level. Each `ReaderContext` is what calls `acquireSearcher` on the searchSupplier (which is ReaderManager). When a shard gets closed it just removes the Reader or PITContext from the `activeReaders` map which means any new search request for same PITContext will fail. For any active search request that will keep working (because of refcounting) and once search is completed then it closes the `ReaderContext` reducing its refcount, which will then close the `searchSupplier` thus decrementing the refCount on the files of the Directory making it eligible for GC.

So I think we should not do anything as part of `close` here for composite directory as well other than updating the state of the directory like `isOpen=false`. This state then needs to be checked in all the methods of this directory specially for IndexOutput and IndexInput providers such that once a directory is closed it doesn't allow any new output/input creation to happen. Also doesn't allow any retrieval of IndexInput from FileCache.